ext/session/tests/gh10807.phpt

```diff
@@ -0,0 +1,83 @@
+--TEST--
+GH-10807 (session_regenerate_id with custom handler and use_strict_mode generates three new session ids)
+--EXTENSIONS--
+session
+--INI--
+html_errors=0
+session.save_handler=files
+session.use_strict_mode=1
+--FILE--
+<?php
+
+/**
+ * This class will return a number for the session id, incrementing each time a new
+ * session is generated
+ */
+class Bug10807SessionHandler implements SessionHandlerInterface, SessionIdInterface {
+
+	private int $fail_validation;
+	private int $sid_counter;
+
+	public function __construct() {
+		$this->fail_validation = 0;
+		$this->sid_counter = 1;
+
+		session_set_save_handler($this, FALSE);
+	}
+
+	public function setFailValidation(int $fail) {
+		$this->fail_validation = $fail;
+	}
+
+	public function create_sid() : string {
+		return strval($this->sid_counter++);
+	}
+
+	public function validateId(string $id) : bool {
+
+		if ($this->fail_validation > 0 && intval($id)< $this->fail_validation) {
+			return FALSE;
+		}
+
+		return TRUE;
+
+	}
```

Comment:
The logic of this is nonsensical.

Validations about the "shape" of the ID *must* happen in the ``create_sid()`` method.

The *purpose* of the ``validateId()`` method is **only** to *validate* that the given ID exists in the data storage.

Please read the Limitation and Compatibility sections of the RFC that introduced those features, which describes the behaviour of this method: https://wiki.php.net/rfc/strict_sessions#limitation

The rationale as to *why* uninitialized sessions should be discarded is explained in the RFC [here](https://wiki.php.net/rfc/strict_sessions#reason_why_the_validation_key_is_required)